File: common/src/main/java/au/com/codeka/warworlds/common/net/PacketEncoder.java
package au.com.codeka.warworlds.common.net;

import au.com.codeka.warworlds.common.Log;
import au.com.codeka.warworlds.common.proto.Packet;
import java.io.IOException;
import java.io.OutputStream;
import javax.annotation.Nullable;
import okio.BufferedSink;
import okio.Okio;

/**
 * Encodes {@link Packet}s onto a stream so that {@link PacketDecoder} can decode them.
 */
public class PacketEncoder {
  private final static Log log = new Log("PacketEncoder");

  public interface PacketHandler {
    void onPacket(Packet packet, int encodedSize);
  }

  private final BufferedSink sink;
  private final Object lock = new Object();
  @Nullable private PacketHandler handler;

  public PacketEncoder(OutputStream outs) {
    this(outs, null);
  }

  public PacketEncoder(OutputStream outs, @Nullable PacketHandler handler) {
    this.sink = Okio.buffer(Okio.sink(outs));
    this.handler = handler;
  }

  public void setPacketHandler(@Nullable PacketHandler handler) {
    this.handler = handler;
  }

  public void send(Packet packet) throws IOException {
    byte[] bytes;
    synchronized (lock) {
      int flags = PacketFlags.NONE;

      bytes = packet.encode();
      byte[] compressed = GzipHelper.compress(bytes);
      if (compressed != null && compressed.length < bytes.length) {
        flags |= PacketFlags.COMPRESSED;
        bytes = compressed;
      }

      sink.writeIntLe(bytes.length);
      sink.writeIntLe(flags);
      sink.write(bytes);
      sink.emit();
    }

    if (handler != null) {
      handler.onPacket(packet, bytes.length);
    }
  }
}

File: common/src/main/java/au/com/codeka/warworlds/common/net/PacketEncoder.kt
package au.com.codeka.warworlds.common.net

import au.com.codeka.warworlds.common.proto.Packet
import okio.BufferedSink
import okio.buffer
import okio.sink
import java.io.OutputStream

/** Encodes [Packet]s onto a stream so that [PacketDecoder] can decode them. */
class PacketEncoder constructor(outs: OutputStream, private var handler: PacketHandler) {
  interface PacketHandler {
    fun onPacket(packet: Packet, encodedSize: Int)
  }

  private val sink: BufferedSink = outs.sink().buffer()
  private val lock = Any()

  fun setPacketHandler(handler: PacketHandler) {
    this.handler = handler
  }

  fun send(packet: Packet) {
    var bytes: ByteArray
    synchronized(lock) {
      var flags = PacketFlags.NONE
      bytes = packet.encode()
      val compressed = GzipHelper.compress(bytes)
      if (compressed != null && compressed.size < bytes.size) {
        flags = flags or PacketFlags.COMPRESSED
        bytes = compressed
      }
      sink.writeIntLe(bytes.size)
      sink.writeIntLe(flags)
      sink.write(bytes)
      sink.emit()
    }
    handler.onPacket(packet, bytes.size)
  }
}
